lib/oban/telemetry.ex

```diff
@@ -158,4 +158,19 @@ defmodule Oban.Telemetry do
       |> Jason.encode!()
     end)
   end
+
+  @doc """
+  Measure the time it takes to run a function and publish corresponding telemetry events.
+  """
+  @spec span(name :: atom(), fun :: (() -> term()), metadata :: map()) :: term()
+  def span(name, fun, metadata \\ %{}) do
+    start_mono = System.monotonic_time(:microsecond)
+    result = fun.()
+
+    :telemetry.execute([:oban, name], %{duration: duration(start_mono)}, metadata)
```

Comment:
If y'all want to add an identifier, we highly suggest putting it into metadata and not as part of the event name. We've come to realize this pattern is causing us a lot of issues. My suggestion for the jobs is to just use `[:oban, :job, :start | :stop | :exception]` for your events.

Reply:
Noted. The plan was to introduce span events just like that after everything was nailed down.

This PR is specifically around emitting other non-job related events, namely dispatching and pruning. Are you suggesting that those events shouldn’t be namespaced?

Reply:
Correct. Event names should be static and anything dynamic moved to a property in the metadata.

Reply:
Ok. To try and clarify a few things.

Currently, _before_ this PR there are five telemetry events:

- `[:oban, :started]`
- `[:oban, :success]`
- `[:oban, :failure]`
- `[:oban, :trip_circuit]`
- `[:oban, :open_circuit]`

That isn't ideal and for those existing events and we want to transition to these events _after_ this PR:

- `[:oban, :job, :start | :stop | :exception]` (with existing metadata)
- `[:oban, :circuit]` (with open/closed as metadata)

For _this_ PR we are looking to add some new events:

- `[:oban, :dispatch, :start | :stop | :exception]` (for producer dispatch)
- `[:oban, :deschedule, :start | :stop | :exception]` (for producer descheduling)
- `[:oban, :prune, :start | :stop | :exception]` (for job pruning)

You're proposing that _instead_ of those events, we introduce a single new event like:

- `[:oban, :query, :start | :stop | :exception]` (with dispatch, deschedule, prune as part of the metadata)

Did I summarize things correctly?

Reply:
Apologies. I thought the name was the name of the job. The _this_ PR ones are 👍 The `:circuit` ones look like they might fit the pattern better with `[:oban, :circuit, :trip | :open]`. Thoughts?

Reply:
Using `[:oban, :circuit, :trip | :open]` looks very appropriate to me. 👍